### PR TITLE
feat(operator): plain kafkasql support

### PR DESCRIPTION
**Pull request by jsenko**

### CODE REVIEW

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/ITBase.java [hunk 1]
```diff
@@ -54,7 +54,7 @@ public enum OperatorDeployment {
     private static Operator operator;
 
     @BeforeAll
-    public static void before() throws Exception {
+    public static void beforeAll() throws Exception {
```
Comment (andreaTP) [inline operator/controller/src/test/java/io/apicurio/registry/operator/it/ITBase.java:57]: nitpick: the operator codebase is still early and I don't mind those changes, going forward let's try to minimize those unrelated changes, and, maybe pack them in a "chore" separate PR.
Two reasons:
- git history -> is easier to follow the history of changes when they are done in isolation
- review -> it keeps things better scoped

File: operator/controller/src/main/resources/k8s/examples/kafkasql/plain/ephemeral.kafka.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+apiVersion: kafka.strimzi.io/v1beta2
```
Comment (andreaTP) [inline operator/controller/src/main/resources/k8s/examples/kafkasql/plain/ephemeral.kafka.yaml:1]: Should those `examples` resources stay in the `test` scope?
Do we need them at runtime in the Operator?

File: operator/controller/src/main/java/io/apicurio/registry/operator/feat/KafkaSql.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import io.apicurio.registry.operator.api.v1.ApicurioRegistry3;
+import io.fabric8.kubernetes.api.model.EnvVar;
+import io.fabric8.kubernetes.api.model.EnvVarBuilder;
+
+import java.util.Map;
+
+import static io.apicurio.registry.operator.resource.app.AppDeploymentResource.addEnvVar;
+import static io.apicurio.registry.operator.utils.Utils.isBlank;
+
+public class KafkaSql {
```
Comment (andreaTP) [inline operator/controller/src/main/java/io/apicurio/registry/operator/feat/KafkaSql.java:12]: This separation is clear and keeps a good readability, happy to follow this style if you prefer it for the Sql datasource :+1:

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        applyStrimziResources(false);
+    }
+
+    @Test
+    void testKafkaSQLPlain() {
+        var kafka = deserialize("/k8s/examples/kafkasql/plain/ephemeral.kafka.yaml", Kafka.class);
+        client.resource(kafka).create();
+
+        await().ignoreExceptions().until(() -> {
+            var status = client.resources(Kafka.class).inNamespace(namespace).withName("my-cluster").get()
```
Comment (andreaTP) [inline operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java:43]: Can we have a check that is not dependent on the Strimzi CR?
E.g. something that should work without code changes also for plain Kafka and RedPanda.

File: operator/controller/pom.xml [hunk 5]
```diff
@@ -89,6 +89,12 @@
       <artifactId>lombok</artifactId>
       <scope>provided</scope>
     </dependency>
+    <dependency>
+      <groupId>io.strimzi</groupId>
```
Comment (andreaTP) [inline operator/controller/pom.xml:93]: Ideally we should be able to not tight the setup to Strimzi but keep it more generic.

File: operator/model/src/main/java/io/apicurio/registry/operator/api/v1/ApicurioRegistry3SpecKafkaSql.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.fasterxml.jackson.annotation.*;
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.fasterxml.jackson.databind.JsonDeserializer.None;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import io.fabric8.kubernetes.api.model.KubernetesResource;
+import lombok.*;
+
+@JsonInclude(Include.NON_NULL)
+@JsonPropertyOrder({ "env", "host" })
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/model/src/main/java/io/apicurio/registry/operator/api/v1/ApicurioRegistry3SpecKafkaSql.java]: Fixed

File: operator/model/src/main/java/io/apicurio/registry/operator/api/v1/ApicurioRegistry3SpecApp.java [hunk 7]
```diff
@@ -23,6 +23,18 @@
...
+    @JsonPropertyDescription("""
+            Configuration of Apicurio Registry SQL storage.""")
+    @JsonSetter(nulls = Nulls.SKIP)
+    private Sql sql;
+
+    @JsonProperty("kafkasql")
+    @JsonPropertyDescription("""
+            Configure KafkaSQL storage.""")
+    @JsonSetter(nulls = Nulls.SKIP)
+    private ApicurioRegistry3SpecKafkaSql kafkasql = new ApicurioRegistry3SpecKafkaSql();
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/model/src/main/java/io/apicurio/registry/operator/api/v1/ApicurioRegistry3SpecApp.java]: Fixed. So the rule is: default value for collections, `null` for objects? If so we need to update `ApicurioRegistry3Spec`.

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @AfterAll
+    public static void afterAll() throws Exception {
+        applyStrimziResources(true);
+        ITBase.afterAll();
+    }
+
+    private static void applyStrimziResources(boolean delete) {
+        String text = load("/k8s/examples/kafkasql/strimzi-cluster-operator-0.43.0.yaml");
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java]: Done. However, I would prefer downloading a specific version, since we're using Strimzi Java model.

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+            assertThat(client.pods().inNamespace(namespace).withName(podName).getLog())
+                    .contains("Using Kafka-SQL artifactStore");
+
+            return true;
+        });
+    }
+
+    @AfterAll
+    public static void afterAll() throws Exception {
+        applyStrimziResources(true);
```
[the remark this answers is not in the document]
Comment (jsenko) [reply operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java]: Reverted. I consider this fine, since it is often easier to fix a small thing immediately if you notice it, rather than trying to remember and create micro PRs. Do you consider it a good practice to track things like these in issues? If so I'll create an issue for this.

File: operator/controller/src/main/resources/k8s/examples/kafkasql/plain/ephemeral.kafka.yaml (hunk 2 above)
Comment (jsenko) [reply operator/controller/src/main/resources/k8s/examples/kafkasql/plain/ephemeral.kafka.yaml:1]: Not necessarily in runtime, but my thinking is that we can use those examples in documentation for example, while making sure they are tested. If you have a suggestion on how to better organize this, let me know.

File: operator/controller/src/main/java/io/apicurio/registry/operator/feat/KafkaSql.java (hunk 3 above)
Comment (jsenko) [reply operator/controller/src/main/java/io/apicurio/registry/operator/feat/KafkaSql.java:12]: yes, thanks!

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/ITBase.java (hunk 1 above)
Comment (andreaTP) [reply operator/controller/src/test/java/io/apicurio/registry/operator/it/ITBase.java]: no need to process for this, just open a PR changing just this bit and we fast-track it.

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java]: > we're using Strimzi Java model

that's something we should aim to avoid. No need from our perspective to bind to Strimzi.

File: operator/controller/src/main/resources/k8s/examples/kafkasql/plain/ephemeral.kafka.yaml (hunk 2 above)
Comment (andreaTP) [reply operator/controller/src/main/resources/k8s/examples/kafkasql/plain/ephemeral.kafka.yaml:1]: Consider the `test/resources` folder, so that they are not going to be bundled with the final JARs

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java (hunk 4 above)
Comment (andreaTP) [reply operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java:44]: We can get the status of the `ReplicaSet` of the Kafka pods, all the checks are, anyhow happening with retries and it should be more than enough to have a stable situation.

**Review by andreaTP — CHANGES_REQUESTED** [review]
This PR looks in great shape now, thanks!

The last bit is to remove the dependency on Strimzi, I can help doing it if required, just let me know :+1:

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java (hunk 4 above)
Comment (jsenko) [reply operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java:44]: Strimzi does not use `ReplicaSet` btw.

**Comment by jsenko** [comment]
> This PR looks in great shape now, thanks!
> 
> The last bit is to remove the dependency on Strimzi, I can help doing it if required, just let me know 👍

Done.

File: operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @Test
+    void testKafkaSQLPlain() {
+        client.load(getClass().getResourceAsStream("/k8s/examples/kafkasql/plain/ephemeral.kafka.yaml"))
+                .create();
+        final var clusterNAme = "my-cluster";
+
+        await().ignoreExceptions().untilAsserted(() ->
+        // Strimzi uses StrimziPodSet instead of ReplicaSet, so we have to check pods
```
Comment (andreaTP) [inline operator/controller/src/test/java/io/apicurio/registry/operator/it/KafkaSQLITTest.java:41]: :+1:

**Review by andreaTP — APPROVED** [review]
Good to go, nice job!
